File: Backend/servers/14 moderation/src/middleware/authorization.py
from fastapi import Request, HTTPException
from src.queries.user.get import get_admin_status


def is_admin(request: Request):
    if not hasattr(request.state, "user_uuid"):
        raise HTTPException(status_code=500, detail="User not authenticated properly")

    user_uuid = request.state.user_uuid
    admin_status = get_admin_status(user_uuid)
    match admin_status:
        case False:
            raise HTTPException(status_code=403, detail="Not an admin")
        case None:
            raise HTTPException(status_code=500, detail="Server error")

File: Backend/servers/15 admin/middleware/authorization.go
// Authorization middleware
package middleware

import (
	"net/http"
	"github.com/gin-gonic/gin"
    "history/queries"
)

func AuthorizationMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
        userUUID, exists := c.Get("user_uuid")
        if !exists {
            c.JSON(
                http.StatusUnauthorized,
                gin.H{"error": "User not authenticated properly"},
            )
            return
        }

        _, ok := userUUID.(string)
        if !ok {
            c.JSON(
                http.StatusInternalServerError,
                gin.H{"error": "Invalid user ID type"},
            )
            return
        }

        admin_status, err := queries.get_admin_status(userUUID)
        if err != nil {
            c.JSON(
                http.StatusInternalServerError,
                gin.H{"error": "Failed to check admin status"},
            )
            c.Abort()
            return
        }
        if !admin_status {
            c.JSON(
                http.StatusForbidden,
                gin.H{"error": "Not an admin"},
            )
        }

        c.Next()
    }
}

